falcon/media/multipart.py

```diff
@@ -0,0 +1,402 @@
+
+            _, params = self._content_disposition
+            self._name = params.get('name')
+
+        return self._name
+
+    @property
+    def media(self):
+        if self._media is None:
+            handler = self._parse_options.media_handlers.find_by_media_type(
+                self.content_type, 'text/plain')
+            self._media = handler.deserialize(
+                self.stream, self.content_type, None)
+
+        return self._media
+
+
+class MultipartForm:
+
+    def __init__(self, stream, boundary, content_length, parse_options):
+        # NOTE(vytas): More lenient check whether the provided stream is not
+        #   already an instance of BufferedStream.
+        # This approach makes testing both the Cythonized and pure-Python
+        #   streams easier within the same test/benchmark suite.
+        if not hasattr(stream, 'read_until'):
+            if isinstance(stream, request_helpers.BoundedStream):
+                stream = BufferedStream(stream.stream.read, content_length)
+            else:
+                stream = BufferedStream(stream.read, content_length)
+
+        self._stream = stream
+        self._boundary = boundary
+        # NOTE(vytas): Here self._dash_boundary is not prepended with CRLF
+        #   (yet) for parsing the prologue. The CRLF will be prepended later to
+        #   construct the inter-part delimiter as per the RFC 7578, section 4.1
+        #   (see the note below).
+        self._dash_boundary = b'--' + boundary
+        self._parse_options = parse_options
+
+    def __iter__(self):
+        delimiter = self._dash_boundary
+        stream = self._stream
+        max_headers_size = self._parse_options.max_body_part_headers_size
+        remaining_part_count = self._parse_options.max_body_part_count
+
+        while True:
+            # NOTE(vytas): Either exhaust the unused stream part, or skip
+            #   the prologue.
+            stream.pipe_until(delimiter)
+            stream.read(len(delimiter))
+
+            if not delimiter.startswith(CRLF):
+                # NOTE(vytas): RFC 7578, section 4.1.
+                #   As with other multipart types, the parts are delimited with
+                #   a boundary delimiter, constructed using CRLF, "--", and the
+                #   value of the "boundary" parameter.
+                delimiter = CRLF + delimiter
+
+            separator = stream.read_until(CRLF, 2, MultipartParseError)
+            if separator == b'--':
+                if stream.peek(2) != CRLF:
+                    raise MultipartParseError('unexpected form epilogue')
+                break
+            elif separator:
+                raise MultipartParseError('unexpected form structure')
+
+            headers = {}
+            headers_block = stream.read_until(_CRLF_CRLF, max_headers_size,
+                                              MultipartParseError)
+            stream.read(4)
```

Comment:
Looks like this consumes the double-CRLF, but it might be nice to have an inline comment to that effect.

Reply:
Should `read_until()` support an option to consume the delimeter?

Reply:
> Should `read_until()` support an option to consume the delimeter (sic)?
That makes sense :+1: I'll check if it can be easily implemented.

Reply:
I guess this is not critical. It just seems like this is a common pattern (read_until followed by consuming the delimiter), hence the suggestion.

Reply:
I do like your suggestion, I just haven't distilled the updated design and implementation yet since the edge case behaviour will likely change as a result as well. For the good, I think.